### DISCUSSION TITLE
acknowledgements

### DISCUSSION THREAD

**Discussion by LinTechSo** [vectordotdev/vector#13548] [Q&A] [answered]
Hi. 
would you please help me figure this out?
I am trying to minimize the event loss in my vector pipeline with disk buffers and acknowledgments
 Kafka(source)--->loki(sink)
but my question is that if I just add enabled acknowledgments on Loki sink, this means that if Loki doesn't answer the requests, the vector sink buffer doesn't block and my vector consume events again after Loki becomes up again. it correct?
or should I add disk buffer anyway in the case of restarting the vector or the time that vector experiencing OOMKIll for example?

**Comment by spencergilbert** [comment, marked as answer]
Hi @LinTechSo - we recommend using blocking memory buffers and end-to-end acknowledgements when you have a robust pub-sub system as your source ([ref](https://vector.dev/docs/setup/going-to-prod/arch/aggregator/#advanced)). In this case until Loki responds positively for the receipt of a given event the commit log for Kafka won't be advanced and no messages will be dropped.

**Reply by LinTechSo** [reply]
Thank you so much for your help.
but what is the point of a disk buffer in these cases? 
Acknowledgments become enough if there was no Kafka?

**Reply by spencergilbert** [reply]
Disk buffers provide a durable store for your events if your upstream source _isn't_ durable. Kafka will guarantee your events are stored and replicated (protected against loss) as well as provide a commit offset (ensuring consumers don't skip any events). With end-to-end acknowledgements _plus_ Kafka, Vector shifts the durability concerns to Kafka.

If you didn't have Kafka as a source and the source _couldn't_ provide those guarantees you would want to use a disk buffer to ensure that the source (which is _less_ robust/durable than Vector) passes responsibility for protecting events to Vector and it's persistent disk buffer.

**Reply by LinTechSo** [reply]
Thanks a lot.
I'm sorry my last question is if we're going to use disk buffer we should add PVC to our vector instance in the Kubernetes cluster. is it correct?

**Reply by spencergilbert** [reply]
Yes - would definitely recommend a PVC for Kubernetes based deployments.
